mmaprototype: improve error handling for constraint normalization

wenyihu6 opened the pull request:
Epic: CRDB-55052 
Release note: none

---
**mmaprototype: improve error handling for constraint normalization**

Previously, span config normalization could panic or fatal on what were
considered developer errors. If a buggy SpanConfig triggered this path, all
nodes would crash and repeatedly crash again on restart. We should never fatal
on user-provided input, even when the underlying bug is ours.

This change replaces panics and fatals with:

- If span config normalization returns an error, leave the conf field empty so
failures propagate without crashing.
- If range constraint analysis fails, leave the constraints empty.
- Add nil checks on results from ensureAnalyzedConstraints to gracefully skip
  ranges with constraints == nil (including conf == nil).

This commit fixes it so that ranges with invalid configs are logged with
warnings and skipped during rebalancing instead of causing node crashes.

---
**mmaprototype: add optional tracing to store-leaseholder-msg in tests**

To support future testing, this commit adds a `trace` argument to the
store-leaseholder-msg datadriven command. When specified, the command returns
the recording span output, making it easier to debug span config processing in
tests. Note that this is currently left unused. Future commits will use it.

---
**mmaprototype: add test for skipping ranges with invalid span configs**

This commit adds skip_range_invalid_config.txt to demonstrate that ranges with
invalid span configs are now handled gracefully.

---
**mmaprototype: add metric for failed normalization**

This commit adds SpanConfigNormalizationError which helps track ranges where
span config normalization encounters an error. It helps operators identify
problematic span configurations in production without requiring log analysis.

Note: this includes cases where best-effort normalization may have succeeded
(producing a usable config despite the error) and does not include cases where
constraint analysis fails.

Resolves: https://github.com/cockroachdb/cockroach/issues/159531

---
**mmaprototype: rate-limit `no constraints analyzed` warning**

wenyihu6 committed:
mmaprototype: improve error handling for constraint normalization

Previously, span config normalization could panic or fatal on what were
considered developer errors. If a buggy SpanConfig triggered this path, all
nodes would crash and repeatedly crash again on restart. We should never fatal
on user-provided input, even when the underlying bug is ours.

This change replaces panics and fatals with:

- If span config normalization returns an error, leave the conf field empty so
failures propagate without crashing.
- If range constraint analysis fails, leave the constraints empty.
- Add nil checks on results from ensureAnalyzedConstraints to gracefully skip
  ranges with constraints == nil (including conf == nil).

This commit fixes it so that ranges with invalid configs are logged with
warnings and skipped during rebalancing instead of causing node crashes.

wenyihu6 committed:
mmaprototype: add test for skipping ranges with invalid span configs

This commit adds skip_range_invalid_config.txt to demonstrate that ranges with
invalid span configs are now handled gracefully.

wenyihu6 committed:
mmaprototype: add optional tracing to store-leaseholder-msg in tests

To support future testing, this commit adds a `trace` argument to the
store-leaseholder-msg datadriven command. When specified, the command returns
the recording span output, making it easier to debug span config processing in
tests. Note that this is currently left unused. Future commits will use it.

wenyihu6 committed:
mmaprototype: add metric for failed normalization

This commit adds SpanConfigNormalizationError which helps track ranges where
span config normalization encounters an error. It helps operators identify
problematic span configurations in production without requiring log analysis.

Note: this includes cases where best-effort normalization may have succeeded
(producing a usable config despite the error) and does not include cases where
constraint analysis fails.

Resolves: https://github.com/cockroachdb/cockroach/issues/159531

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/159653)
<!-- Reviewable:end -->

tbg reviewed (commented):
Quick, superficial review. Looks good - left some ideas, but I'll defer to you and @sumeerbhola to settle on something pragmatic.

> @tbg reviewed 5 of 5 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-:-OgfZfaL9bXFTruKaABn:beaopma)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1085 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfZXjrE3o_iXrqc5Gy:-OgfZXjrE3o_iXrqc5Gz:b-oi8gdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b6a8452d2b7d7988c0dbd49fe89c603ade42657/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1085)):*
> ```Go
> 	replicas []StoreIDAndReplicaState
> 	// conf may be nil or non-nil depending on the type of normalization error:
> 	// - Hard error: a developer error or the input config is invalid. Results in
> ```

adding to my earlier comment, it does make sense to distinguish hard and soft errors. So maybe a counter for (hard+soft), and one for only hard.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfYWtIDOyacjDKPlGx:-OgfYWtIDOyacjDKPlGy:bgmtttz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8179d2550849a0b55cc20c590895a56ce8f58464/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L49)):*
> ```Go
> 	// may have succeeded (producing a usable config despite the error). This
> 	// metric does not include cases constraint analysis fails.
> 	// TODO(wenyihu6): add ConstraintAnalysisSkipped for those.
> ```

For easier monitoring, it would be good to have one metric that is incremented whenever *anything* goes wrong. We could then have additional metrics that get incremented only for some errors (i.e. distinguish constraint analysis errors vs normalization errors), but I'm not sure we need that much. Once we see the metric increase, we'd look at the logs and figure out what exactly is failing. So consider a more catch-all metric here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfYmJa6WLM8vmZ9Khb:-OgfYmJa6WLM8vmZ9Khc:b-ldwzqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8179d2550849a0b55cc20c590895a56ce8f58464/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L167)):*
> ```Go
> 			"(producing a usable config despite the error).",
> 		Measurement: "Range",
> 		Unit:        metric.Unit_COUNT,
> ```

Would this make more sense as a Gauge, too? So that in each pass we're keeping track of how many times we ran into a bad zone config (vs. tracking the rate at which we do, which is a function of how often we run). Ideally we'd even track the number of distinct configs with issues (rather than the number of ranges we hit an error on, which on many of them will be due to the same config). But - not worth getting overly bogged down in the details here, I think. I am fine with the counter, but if we can EASILY do it, a gauge showing the number of distinct configs we ran into trouble on would be best. If I squint, this is easy to achieve by iterating through all configs once at the beginning of the rebalance loop and deciding on what the current gauge value is there and then.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 4 of 5 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-:-Oghp3UGED4DgevAa9Vp:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfYWtIDOyacjDKPlGx:-OghmdXL9lFBiEpZsO9N:bv831kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8179d2550849a0b55cc20c590895a56ce8f58464/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L49)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

For easier monitoring, it would be good to have one metric that is incremented whenever *anything* goes wrong. We could then have additional metrics that get incremented only for some errors (i.e. distinguish constraint analysis errors vs normalization errors), but I'm not sure we need that much. Once we see the metric increase, we'd look at the logs and figure out what exactly is failing. So consider a more catch-all metric here.
</blockquote></details>

Perhaps we should split into two categories of serious errors that need user attention:
- errors that only affect some ranges, and most MMA behavior can continue fine
- errors that get caught at the top-level (the catching a panic at the top level we discussed today), and *may* have prevented most MMA behavior.
These two could use static labels.

----

Actually, perhaps breaking this into multiple metrics may be too much. I don't have a strong opinion either way.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfYmJa6WLM8vmZ9Khb:-OghoQvl3ldeKoImyu64:b3qeuqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8179d2550849a0b55cc20c590895a56ce8f58464/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L167)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Would this make more sense as a Gauge, too? So that in each pass we're keeping track of how many times we ran into a bad zone config (vs. tracking the rate at which we do, which is a function of how often we run). Ideally we'd even track the number of distinct configs with issues (rather than the number of ranges we hit an error on, which on many of them will be due to the same config). But - not worth getting overly bogged down in the details here, I think. I am fine with the counter, but if we can EASILY do it, a gauge showing the number of distinct configs we ran into trouble on would be best. If I squint, this is easy to achieve by iterating through all configs once at the beginning of the rebalance loop and deciding on what the current gauge value is there and then.
</blockquote></details>

+1 to a gauge and it could be a followup PR. We already iterate through all ranges when processing the StoreLeaseholderMsg and we can maintain the gauge there.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OghfCJ60QTEloHyKduj:-OghfCJ60QTEloHyKduk:bz3nh5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L473)):*
> ```Go
> 		re.ensureAnalyzedConstraints(rstate)
> 		if rstate.constraints == nil {
> 			log.KvDistribution.Warningf(ctx,
> ```

This should be at a reduced frequency, say every 1min.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OghfGfQEkMR47OW0JtT:-OghfGfQEkMR47OW0JtU:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L677)):*
> ```Go
> 		if rstate.constraints == nil {
> 			log.KvDistribution.Warningf(ctx,
> 				"skipping r%d: no constraints analyzed (conf=%v replicas=%v)",
> ```

ditto
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OghdN69Bum6xFgRFKVb:-OghdN69Bum6xFgRFKVc:b7jjmp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1088)):*
> ```Go
> 	// best-effort result and the error is logged but not propagated here.
> 	//
> 	// If conf is nil, range analysis is skipped, and thus constraints is nil.
> ```

nit: ... thus constraints is always nil.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1090 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-Oghd_xe9Znx7ljIzhj_:-Oghd_xe9Znx7ljIzhja:b-w7aicw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1090)):*
> ```Go
> 	// If conf is nil, range analysis is skipped, and thus constraints is nil.
> 	// This range will be skipped from rebalancing. If conf is non-nil (even if it
> 	// was a best-effort result), range analysis is performed.
> ```

I have a mild preference for a less verbose comment.
// conf is nil iff basic span config normalization failed (e.g., invalid
// config). NB: if only structural normalization failed, conf is non-nil.
// When conf is nil, range analysis is skipped, and thus rangeAnalyzedConstraints is always nil, and the range is skipped for rebalancing etc.

That is, I am not sure we need comments about what is a best-effort results and what observability we have for that, here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgheaN-AhI28jUaEyZG:-OgheaN-AhI28jUaEyZH:b52ofyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1106)):*
> ```Go
> 	// If non-nil, it is up-to-date. Typically, non-nil for a range that has no
> 	// pendingChanges and is not satisfying some constraint, since we don't want
> 	// to repeat the analysis work every time we consider it.
> ```

We should hoist the new text here since we already discuss nil-ness here. I think it would suffice to say
// Always nil if conf is nil (since conf is needed as input in constructing these analyzed constraints).


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: rate-limit `no constraints analyzed` warning

wenyihu6 reviewed (commented):
TFTR! I’ll put up a follow-up PR to address the feedback around using gauge and more granular metrics. Will also hold off on the backport until the follow up PR.

> @wenyihu6 made 9 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-:-Oglgqd_D5L9ClFzjx2r:b3fqsq8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @tbg).

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OghdN69Bum6xFgRFKVb:-OglgSFN7WyXmr5tEvd4:b-qj92pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1088)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: ... thus constraints is always nil.
</blockquote></details>

Reworded.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1090 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-Oghd_xe9Znx7ljIzhj_:-OglgUdvALpCgorssn5W:b-qj92pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1090)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I have a mild preference for a less verbose comment.
// conf is nil iff basic span config normalization failed (e.g., invalid
// config). NB: if only structural normalization failed, conf is non-nil.
// When conf is nil, range analysis is skipped, and thus rangeAnalyzedConstraints is always nil, and the range is skipped for rebalancing etc.

That is, I am not sure we need comments about what is a best-effort results and what observability we have for that, here.
</blockquote></details>

Reworded.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgheaN-AhI28jUaEyZG:-OglgVhB1Vqs6-Xu5pK0:b-qj92pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1106)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We should hoist the new text here since we already discuss nil-ness here. I think it would suffice to say
// Always nil if conf is nil (since conf is needed as input in constructing these analyzed constraints).
</blockquote></details>

Reworded.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go` line 1085 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfZXjrE3o_iXrqc5Gy:-OglfyvZ6LCZL9qP0fJR:b-8hd7gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b6a8452d2b7d7988c0dbd49fe89c603ade42657/pkg/kv/kvserver/allocator/mmaprototype/cluster_state.go#L1085)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

adding to my earlier comment, it does make sense to distinguish hard and soft errors. So maybe a counter for (hard+soft), and one for only hard.
</blockquote></details>

Agreed. I held off on the new metrics due to conflicts with PR #159556, which has been merged now. I’ll put up a follow-up PR to address this.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OghfCJ60QTEloHyKduj:-OglgXJB6vtbcIbpB3FN:b-e8qrep) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L473)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This should be at a reduced frequency, say every 1min.
</blockquote></details>

Added the rate limit.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OghfGfQEkMR47OW0JtT:-OglgaGz5pf9NusXd00j:b-e8qrep) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dce15fb86a4dea4df60133de1791e6840da1d2c/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L677)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Added the rate limit.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfYWtIDOyacjDKPlGx:-Oglgaw46dYcO6nIJFSh:b-blaswl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8179d2550849a0b55cc20c590895a56ce8f58464/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L49)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Perhaps we should split into two categories of serious errors that need user attention:
- errors that only affect some ranges, and most MMA behavior can continue fine
- errors that get caught at the top-level (the catching a panic at the top level we discussed today), and *may* have prevented most MMA behavior.
These two could use static labels.

----

Actually, perhaps breaking this into multiple metrics may be too much. I don't have a strong opinion either way.
</blockquote></details>

Agreed, will address in the follow up pr. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go` line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/159653#-OgfYmJa6WLM8vmZ9Khb:-OglglT09TRGvHPwDsSQ:b-r7od4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8179d2550849a0b55cc20c590895a56ce8f58464/pkg/kv/kvserver/allocator/mmaprototype/mma_metrics.go#L167)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

+1 to a gauge and it could be a followup PR. We already iterate through all ranges when processing the StoreLeaseholderMsg and we can maintain the gauge there.
</blockquote></details>

Good ideas. I will put up a follow up pr. 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: sort stores by id in loadSummaryForAllStores for deterministic output

Previously, processStoreLeaseholderMsgInternal iterated over cs.stores map
directly, which caused non-deterministic logs in datadriven tests. This change
sorts store IDs before iterating to ensure deterministic log output in tests.
Alternatively, we could also pass in a boolean that controls deterministic
output.

Release Notes: None
Epic: None
